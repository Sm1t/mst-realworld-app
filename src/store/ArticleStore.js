import { types, flow } from 'mobx-state-tree';

import { ArticleService } from '../api';

const Author = types.model({
  username: types.string,
  bio: types.maybeNull(types.string),
  image: types.string,
  following: types.boolean,
});

export const Article = types
  .model({
    title: types.string,
    slug: types.string,
    body: types.string,
    createdAt: types.string,
    updatedAt: types.string,
    tagList: types.array(types.string),
    description: types.string,
    author: Author,
    favorited: types.boolean,
    favoritesCount: types.number,
  })
  .actions(self => ({
    setTitle(title) {
      self.title = title;
    },
    setContent(content) {
      self.content = content;
    },
  }));

export const ArticleStore = types
  .model({
    articles: types.optional(types.array(Article), []),
  })
  .actions(self => {
    function add(item) {
      self = self.articles.push(item);
    }

    const loadArticles = flow(function* loadArticles() {
      try {
        const {
          data: { articles },
        } = yield ArticleService.all();

        self.articles = articles;
      } catch (err) {
        console.error('Failed to load articles ', err);
      }
    });

    return {
      add,
      loadArticles,
    };
  });
