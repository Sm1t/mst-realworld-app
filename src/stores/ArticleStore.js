import { types, flow } from 'mobx-state-tree';

import { ArticleService } from '../api';

export const Article = types
  .model({
    title: types.string,
    content: types.string,
  })
  .actions(self => ({
    setTitle(title) {
      self.title = title;
    },
    setContent(content) {
      self.content = content;
    },
  }));

export const ArticleStore = types
  .model({
    articles: types.optional(types.array(Article), []),
  })
  .actions(self => {
    function add(item) {
      self = self.articles.push(item);
    }

    const loadArticles = flow(function* loadArticles() {
      try {
        const {
          data: { articles },
        } = yield ArticleService.all();

        self.articles = articles.map(({ title, body }) => ({
          title,
          content: body,
        }));
      } catch (err) {
        console.error('Failed to load articles ', err);
      }
    });

    return {
      add,
      loadArticles,
    };
  });
