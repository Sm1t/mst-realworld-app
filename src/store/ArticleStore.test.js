import { getSnapshot, onSnapshot } from 'mobx-state-tree';

import { Article, ArticleStore } from './ArticleStore';

const articleData = {
  title: 'Test1',
  slug: 'test1',
  body: 'Longer text',
  createdAt: '2016-12-16T10:57:00.990Z',
  updatedAt: '2016-12-16T10:57:00.990Z',
  tagList: [],
  description: 'Desc',
  author: {
    username: 'memas',
    bio: null,
    image: 'https://static.productionready.io/images/smiley-cyrus.jpg',
    following: false,
  },
  favorited: false,
  favoritesCount: 1,
};

describe('models', () => {
  describe('Article', () => {
    it('should properly create article', () => {
      const article = Article.create(articleData);

      expect(getSnapshot(article)).toMatchSnapshot();
    });

    it('should properly set article fields', () => {
      const article = Article.create(articleData);

      article.setTitle('Lorem');
      article.setContent('Lorem ipsum dolor sit amet');

      expect(getSnapshot(article)).toMatchSnapshot();
    });
  });

  describe('ArticleList', () => {
    it('should create an empty list when items is not passed', () => {
      const articleList = ArticleStore.create();

      expect(getSnapshot(articleList)).toMatchSnapshot();
    });

    it('should create a list of articles', () => {
      const articleList = ArticleStore.create({ items: [articleData] });

      expect(getSnapshot(articleList)).toMatchSnapshot();
    });

    it('should properly add new article', () => {
      const store = ArticleStore.create();
      const states = [];

      onSnapshot(store, snapshot => states.push(snapshot));

      expect(getSnapshot(store)).toMatchSnapshot();
      store.add(articleData);
      expect(states).toMatchSnapshot();
    });
  });
});
