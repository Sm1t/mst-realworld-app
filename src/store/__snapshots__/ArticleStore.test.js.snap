// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models Article should properly create article 1`] = `
Object {
  "author": Object {
    "bio": null,
    "following": false,
    "image": "https://static.productionready.io/images/smiley-cyrus.jpg",
    "username": "memas",
  },
  "body": "Longer text",
  "createdAt": "2016-12-16T10:57:00.990Z",
  "description": "Desc",
  "favorited": false,
  "favoritesCount": 1,
  "slug": "test1",
  "tagList": Array [],
  "title": "Test1",
  "updatedAt": "2016-12-16T10:57:00.990Z",
}
`;

exports[`models Article should properly set article fields 1`] = `
Object {
  "author": Object {
    "bio": null,
    "following": false,
    "image": "https://static.productionready.io/images/smiley-cyrus.jpg",
    "username": "memas",
  },
  "body": "Longer text",
  "createdAt": "2016-12-16T10:57:00.990Z",
  "description": "Desc",
  "favorited": false,
  "favoritesCount": 1,
  "slug": "test1",
  "tagList": Array [],
  "title": "Lorem",
  "updatedAt": "2016-12-16T10:57:00.990Z",
}
`;

exports[`models ArticleList should create a list of articles 1`] = `
Object {
  "articles": Array [],
}
`;

exports[`models ArticleList should create an empty list when items is not passed 1`] = `
Object {
  "articles": Array [],
}
`;

exports[`models ArticleList should properly add new article 1`] = `
Object {
  "articles": Array [],
}
`;

exports[`models ArticleList should properly add new article 2`] = `
Array [
  Object {
    "articles": Array [
      Object {
        "author": Object {
          "bio": null,
          "following": false,
          "image": "https://static.productionready.io/images/smiley-cyrus.jpg",
          "username": "memas",
        },
        "body": "Longer text",
        "createdAt": "2016-12-16T10:57:00.990Z",
        "description": "Desc",
        "favorited": false,
        "favoritesCount": 1,
        "slug": "test1",
        "tagList": Array [],
        "title": "Test1",
        "updatedAt": "2016-12-16T10:57:00.990Z",
      },
    ],
  },
]
`;
